File: javascript/packages/core/src/support/priority.js
/** Order priority and its pricing effect. */

/**
 * How urgently an order is handled.
 *
 * @typedef {"standard" | "rush" | "warranty"} PriorityValue
 */

/** @type {Readonly<Record<string, number>>} */
const SURCHARGE_BP = Object.freeze({
    standard: 0,
    rush: 2500,
    warranty: 0,
});

/** @type {Readonly<Record<string, string>>} */
const LABELS = Object.freeze({
    standard: "Standard",
    rush: "Rush",
    warranty: "Warranty",
});

export const Priority = Object.freeze({
    Standard: /** @type {PriorityValue} */ ("standard"),
    Rush: /** @type {PriorityValue} */ ("rush"),
    Warranty: /** @type {PriorityValue} */ ("warranty"),

    /**
     * Surcharge in basis points applied to the labour subtotal.
     *
     * @param {PriorityValue} priority
     * @returns {number}
     */
    surchargeBp(priority) {
        return SURCHARGE_BP[priority] ?? 0;
    },

    /**
     * @param {PriorityValue} priority
     * @returns {string}
     */
    label(priority) {
        return LABELS[priority] ?? priority;
    },

    /**
     * Rush work jumps the queue; warranty work does not.
     *
     * @param {PriorityValue} priority
     * @returns {boolean}
     */
    isExpedited(priority) {
        return priority === "rush";
    },

    /**
     * Warranty orders are never invoiced to the customer.
     *
     * @param {PriorityValue} priority
     * @returns {boolean}
     */
    isBillable(priority) {
        return priority !== "warranty";
    },
});

File: rust/crates/atelier-core/src/support/priority.rs
//! Order priority and its pricing effect.

use serde::{Deserialize, Serialize};

/// Priority tier chosen at intake.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Priority {
    #[default]
    Standard,
    Rush,
    Warranty,
}

impl Priority {
    /// Surcharge in basis points applied to the labour subtotal.
    pub fn surcharge_bp(&self) -> u32 {
        match self {
            Priority::Standard => 0,
            Priority::Rush => 2_500,
            Priority::Warranty => 0,
        }
    }

    pub fn label(&self) -> &'static str {
        match self {
            Priority::Standard => "Standard",
            Priority::Rush => "Rush",
            Priority::Warranty => "Warranty",
        }
    }

    /// Rush work jumps the queue; warranty work does not.
    pub fn is_expedited(&self) -> bool {
        matches!(self, Priority::Rush)
    }

    /// Warranty orders are never invoiced to the customer.
    pub fn is_billable(&self) -> bool {
        !matches!(self, Priority::Warranty)
    }
}
